### PR TITLE
feat(task): task-local runtime thinking effort state with per-request override (DTE series 2/5)

### PR DESCRIPTION
**Pull request by easonLiangWorldedtech**
## ≤400-line redo of Zoo-Code-Org/Zoo-Code#1338 — DTE series 2/5, unit 3/5

Task-local runtime thinking-effort state on `Task`: the in-memory override
channel, its per-request delivery at all four `createMessage` sites, and the
profile-switch re-capture in `updateApiConfiguration`. Transient state only — nothing is persisted; persistence is the next unit (U4).

### Stack

- PR base: upstream `main` @ `0dbd5846f` (cross-fork PR — head `easonLiangWorldedtech:feat/dte-v2-3-task-runtime-effort`; stack branches live in the fork — no push access to create them here)
- Stack base: `feat/dte-v2-2-per-request-effort` @ `069c34b9a` (U2, upstream PR #1522, new tip — carries U1's post-main-sync head `e89cceddd` and the advanced upstream `main` tip `0dbd5846f` merged in)
- Stack sync: first sync `f97f8d999` merged U2's tip `18f488fa5` (U1 final CR fixes `53f22dc` in); second sync `c5b48aa7e` merged U2's new tip `efbd336e5` (U1's final head `39762bf81` — the last #1521 CR fix — in); third (main) sync `e8c66cfad` (this head) merges U2's post-main-sync tip `069c34b9a` (upstream `main` advanced `0d937c050` → `0dbd5846f`: v3.82.0 release prep #1533, GPT-6 Astra #1506, DeepSeek V4 Flash Vision #1488, the `throwIfAborted` helper + `completePrompt` options regression tests #1288, and the async `Task.dispose()` test-teardown fix #1527). One conflict in `src/core/task/Task.ts` resolved: main's async `dispose(): Promise<void>` restructure (memoized `disposalPromise` + `disposeOnce()`) kept, with U3's 4-line DTE JSDoc above the new `dispose()`; the test's `afterEach` now awaits `task.dispose()` per the repo idiom. Standalone budget 398 at this head (amended to 403 by the CR re-review fix in commit `1edd728c4`, then to 418 by the follow-up CR fix in commit `1087fcbd4`, see below)
- Content source: `feat/dte-trial-all` union `27a2e97df` (tag `dte-legacy/union`), U3 slice
- Related: easonLiangWorldedtech/Zoo-Code#35 (DTE-v2 tracking ledger)

GitHub's displayed diff vs `main` is cumulative over the unmerged lower units
(U1 #1521, U2 #1522); the standalone range below is the review target — the displayed number shrinks as they merge. Merge this PR only after its stack
base PR has merged.

### Budget (plan §2: a+d ≤400 soft target; ≤1000 hard)

`2 files changed, 417 insertions(+), 1 deletion(-)` = **418** — soft target 400 exceeded by 18, CR-driven (see the amendments below); hard cap 1000 ✓
| File | a+d |
|------|-----|
| `src/core/task/Task.ts` | 102+/1− |
| `src/core/task/__tests__/Task.runtime-thinking-effort.test.ts` (new) | 315+ |

**Budget deviation note (plan §2.6).** The plan estimated U3 at ~355
(Task +132/− + tests ~20); those numbers were stale. Measured against the
union, the U3 slice is Task +105/− + a 311-line test file = 417 > 400. Per
§2.6 (no budget bypass), the **dispose boundary group** is split into U4:
the task-end override reset (6 Task lines + 3 DTE JSDoc lines) and its
12-line `describe("dispose")` test block. This matches the plan's own U4 scope
line ("persistence + boundary cases"). U3 keeps the generic 4-line `dispose()`
JSDoc; U4 expands it with the DTE sentence alongside the reset code.

**CR re-review amendment (2026-09-05, review `5120440807`).** The main-sync
head re-review requested the task-end override reset to live in
`disposeOnce()` of this unit (so a retained disposed task never serves a stale
override, and the `dispose()` JSDoc's "resets transient task-local state"
claim is accurate), plus the test `afterEach` teardown to call
`await task.dispose()` unconditionally (the `!task.abort` guard had skipped
disposal of aborted tasks). Both land here in commit `1edd728c4` (+5/−1 ⇒
standalone 398 → 403 — a 3-line overshoot of the soft target, within the 1000
hard cap). Under the split above, U4 loses the reset code it had carried; its
12-line dispose-boundary test block stays with U4 (it now exercises this
unit's reset on the stacked tree).

**CR follow-up fix (2026-09-06, review thread comment `3942121231`).** The
re-review of head `1edd728c4` found that the unconditional field clears in
`disposeOnce()` left the override behind in `apiConfiguration.reasoningEffort`
and the built `api` handler — both are populated by
`setRuntimeThinkingEffort`, so a retained disposed task could still expose the
override through those copies. Commit `1087fcbd4` replaces the three
assignments with the standard clearing call `this.setRuntimeThinkingEffort(undefined)` —
which restores `apiConfiguration`/`api` from `preOverrideReasoningEffort`
(read before it is cleared) and then clears the runtime fields — and adds a
14-line `describe("dispose")` killing-test block asserting the clear and the
restore (Task.ts net +1; standalone 403 → 418 a+d — CR-driven, ≤1000 hard cap).

### Provenance / fidelity

- `Task.ts`: 3-way `git merge-file` — base `39bdfb188` (= `6ea45b36a^`),
  ours = U2 head, theirs = `90b47b053` (the last U3 commit, before the U4
  persistence work). **Zero conflicts.** A whole-file extract was impossible:
  the union's `Task.ts` carries U14-orchestrator and U4/U5 content
  (215+/241− vs U1 head), and per-commit `git apply --3way` of the U3 patches
  fails on upstream base drift.
- Test file: byte-exact `90b47b053` version (311 lines) minus the dispose
  describe (12 lines + separator), plus the 3 mutation-killing assertion
  lines below = 301 lines; the header comment is trimmed to the U3 scope
  ("the task-end reset in dispose()" clause moves with U4).
- U4 content excluded: `taskMetadata.ts` / `history.ts` persistence changes,
  the `describe("history persistence round-trip")` and
  `describe("abortTask final save")` blocks, and the `HistoryItem` import
  (unused in the U3 slice).
- `src/eslint-suppressions.json`: untouched. The union's +21 suppression-count
  deltas vs the stack base are all in files owned by other units
  (`gemini-format.spec.ts` 5→6, `ask-queued-message-drain.spec.ts` 18→32,
  `newTaskTool.spec.ts` 26→31, new `extension.ts` 1) — none U3-owned.

### Out of scope (next units)

- **U4 (DTE-2c):** persist task thinking effort to history items,
  `taskMetadata` merge propagation, and the task-end override reset split out
  above (dispose boundary + its test).
- **U5 (DTE-2d):** the Anthropic `output_config.effort` adaptive envelope.

### Mutation-diff fix (killing assertions, plan L42 — same PR)

The first CI `mutation-diff` run (head `d0b1a3dcd`) reported 2 Survived
`ConditionalExpression` mutants — both on the two ternaries this unit
introduces:

| Location | Surviving variant | Fix (this PR) |
|---|---|---|
| `setRuntimeThinkingEffort` source capture: `effort === undefined ? undefined : source` | replacement: false ⇒ always `source`: a label passed on a clearing call leaks into `source` | the clearing call now carries a label (`setRuntimeThinkingEffort(undefined, "stale-source")`); the existing `source: undefined` assertion then kills the variant |
| `getRuntimeThinkingEffortMetadata`: `effort !== undefined ? { reasoningEffort } : {}` | replacement: true ⇒ always `{ reasoningEffort: <maybe undefined> }` | key-absence assertion `not.toHaveProperty("reasoningEffort")` while unset — `toEqual({})` cannot kill it (toEqual ignores keys whose value is `undefined`); asserted pre-set and post-clear |

The complementary variants (L1683-true, L1721-false) were already killed by
the existing `toBe("test-source")` and `toEqual({ reasoningEffort: "high" })`
assertions.

Local dev-stage gate (skill §5.1) on the latest CR-fix head `1087fcbd4`:
`node scripts/stryker-diff.mjs ci --base 069c34b9a --head 1087fcbd4` — extension **36 changed lines, 26 valid mutants, 26 Killed, 0 Survived / 0 NoCoverage, exit 0** (previous CR-fix head `1edd728c4`: 38 lines, 25/25 Killed).

### Verification (local, head `1087fcbd4`)

- `pnpm --dir src exec eslint --prune-suppressions --max-warnings=0 core/task/Task.ts core/task/__tests__/Task.runtime-thinking-effort.test.ts` — exit 0 (re-run on `1087fcbd4`; no suppression-count change)
- `pnpm check-types` — 11/11 projects
- `pnpm --dir src exec vitest run core/task/__tests__/Task.runtime-thinking-effort.test.ts` — 9/9 (re-run on `1087fcbd4`, 4.16 s; includes the new dispose killing test)
- `git diff --shortstat 069c34b9a HEAD` — 417+/1− = 418 (CR fixes +19/−4 vs `1edd728c4`; soft overshoot documented above, ≤1000 hard)
- mutation-diff gate (dev stage, skill §5.1) on `1087fcbd4` — exit 0, 26/26 Killed (see fix above)

### CODE REVIEW

File: src/core/task/__tests__/Task.runtime-thinking-effort.test.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+			provider: mockProvider as unknown as ClineProvider,
+			apiConfiguration: mockApiConfiguration,
+			startTask: false,
+		})
+	})
+
+	afterEach(async () => {
+		vi.useRealTimers()
+		if (task && !task.abort) {
+			await task.dispose().catch(() => {})
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/task/__tests__/Task.runtime-thinking-effort.test.ts]: Fixed in `1edd728c4`: the `afterEach` teardown now calls and awaits `task.dispose()` whenever `task` exists — the `!task.abort` guard is removed (it had skipped disposal of aborted tasks), and the swallowed disposal-error handling is retained. DTE suite green locally (8/8).

**Comment by easonLiangWorldedtech** [comment]
Addressed CR re-review `5120440807` in commit `1edd728c4`:

1. **`Task.ts` / `disposeOnce()` (the outside-diff finding):** the three DTE fields (`runtimeThinkingEffort`, `runtimeThinkingEffortSource`, `preOverrideReasoningEffort`) are now cleared on disposal, so a retained disposed task never serves a stale override — the `dispose()` JSDoc's “resets transient task-local state” is accurate.
2. **test teardown guard:** fixed in-thread (unconditional `await task.dispose()`).

Standalone budget is now 402+/1− = 403 (3 over the 400 soft target, CR-driven; within the 1000 hard cap) — documented in the Budget section. Local gates on `1edd728c4`: check-types 11/11, vitest 8/8, eslint 0, mutation-diff gate 25/25 Killed (extension 38 lines, exit 0).

File: src/core/task/Task.ts [hunk 2]
```diff
@@ -2617,6 +2706,11 @@ export class Task extends EventEmitter<TaskEvents> implements TaskLike {
 			console.error("Error reverting diff changes:", error)
 		}
 
+		// Clear DTE runtime-effort override state so a retained disposed task never serves a stale override.
+		this.runtimeThinkingEffort = undefined
+		this.runtimeThinkingEffortSource = undefined
+		this.preOverrideReasoningEffort = undefined
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/task/Task.ts]: Fixed in `1087fcbd4`: the dispose-time clear now goes through the standard clearing call - `disposeOnce()` calls `this.setRuntimeThinkingEffort(undefined)` instead of assigning the three fields directly. That call restores `apiConfiguration.reasoningEffort` from `preOverrideReasoningEffort` (read before it is cleared) and rebuilds `this.api` via `buildApiHandler`, so the override no longer lingers in the in-memory `apiConfiguration` copy or the built handler; the three runtime fields are cleared as before.

Regression test added (same file): the `describe("dispose")` block asserts that after `setRuntimeThinkingEffort("xhigh", "source")` + `await task.dispose()`, `getRuntimeThinkingEffort()` is empty, all three runtime fields are `undefined`, and `task.apiConfiguration.reasoningEffort` is restored to the settings-derived value.

Local gates on `1087fcbd4`: check-types 11/11, targeted vitest 9/9 (4.16 s), eslint exit 0 (no suppression-count change), mutation-diff dev-stage gate (base `069c34b9a`) exit 0 - 26 valid mutants, 26 Killed. Standalone budget 403 -> 418 a+d (CR-driven; documented in the PR body, <=1000 hard cap).

**Comment by easonLiangWorldedtech** [comment]
**CR re-review follow-up (head `1edd728c4` -> `1087fcbd4`).**

The re-review of `1edd728c4` raised one new finding (thread comment `3942121231`): the unconditional field clears in `disposeOnce()` left the override behind in `apiConfiguration.reasoningEffort` and the built `api` handler - both populated by `setRuntimeThinkingEffort`, so a retained disposed task could still expose the override through those copies.

Fixed additively in `1087fcbd4`: `disposeOnce()` now calls the standard clearing call `setRuntimeThinkingEffort(undefined)` (restores `apiConfiguration`/`api` from `preOverrideReasoningEffort`, then clears the runtime fields), plus a 14-line `describe("dispose")` killing-test block asserting the clear and the restore.

Gates on `1087fcbd4`: check-types 11/11, targeted vitest 9/9 (4.16 s), eslint exit 0, mutation-diff (dev stage, vs `069c34b9a`) exit 0 - 26 valid mutants, 26 Killed. Standalone 417+/1- = 418 a+d (soft overshoot CR-driven, <=1000 hard). Body updated.
